pulsar-broker/src/main/java/org/apache/pulsar/broker/service/nonpersistent/NonPersistentTopic.java

```diff
@@ -207,9 +207,20 @@ public void addProducer(Producer producer) throws BrokerServiceException {
                 log.debug("[{}] {} Got request to create producer ", topic, producer.getProducerName());
             }
 
-            if (!producers.add(producer)) {
-                throw new NamingException(
-                        "Producer with name '" + producer.getProducerName() + "' is already connected to topic");
+            Producer existProducer = producers.putIfAbsent(producer.getProducerName(), producer);
+            if (existProducer != null) {
+                boolean canOverwrite = false;
+                if (existProducer.equals(producer) && !existProducer.isUserProvidedProducerName()
+                        && !producer.isUserProvidedProducerName() && producer.getEpoch() > existProducer.getEpoch()) {
+                    existProducer.close();
+                    canOverwrite = true;
+                }
+                if (canOverwrite) {
+                    producers.put(producer.getProducerName(), producer);
```

Comment:
use `producers.replace(producer.getProducerName(), existingProducer, producer)` to make sure one can successfully add the producer.